sklearn/model_selection/_search.py

```diff
@@ -320,24 +324,61 @@ def fit_grid_point(X, y, estimator, parameters, train, test, scorer,
     return score, parameters, n_samples_test
 
 
+def _check_param_dist(param_dist):
```

Comment:
If the only reason you've invented this check is to get a set of parameter names, you're almost certainly going about things the wrong way. I don't see the problem with getting parameter names on the fly, and being able to accept an arbitrary object supporting `len` and `iter` to `BaseSearchCV.fit` as we currently do.